### PR TITLE
ISPN-10753 Use StampedLock in StateTransferLockImpl and several other places

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.redhat.com/browse/ISPN-10753

### CODE REVIEW

**Comment by wburns** [comment]
The changes look okay. However, I worry that I am not familiar with all the changes to be sure there isn't any reentrancy issues as StampedLock doesn't allow reentrancy.

**Comment by wburns** [comment]
I reran the build, but it seemed to fail again. Guessing the changes broke something @danberindei

**Comment by pruivo** [comment]
```
        at java.util.concurrent.locks.LockSupport.park(java.base@11.0.11/LockSupport.java:194)
        at java.util.concurrent.locks.StampedLock.acquireRead(java.base@11.0.11/StampedLock.java:1502)
        at java.util.concurrent.locks.StampedLock.readLock(java.base@11.0.11/StampedLock.java:539)
        at org.infinispan.client.hotrod.impl.transport.netty.ChannelFactory.getServers(ChannelFactory.java:403)
        at org.infinispan.client.hotrod.impl.transport.netty.ChannelFactory.getNextServer(ChannelFactory.java:281)
        at org.infinispan.client.hotrod.impl.transport.netty.ChannelFactory.fetchChannelAndInvoke(ChannelFactory.java:264)
```
`ChannelFactory#getNextServer()` already acquires the lock and `ChannelFactory#getServers()` tries to re-acquire it.

**Comment by danberindei** [comment]
Sorry @wburns @pruivo, I'm not sure how I missed the problems with the `ChannelFactory` commit. I tried to eliminate the reentrancy yesterday, but I realized I really don't want to *just* fix the reentrancy in `ChannelFactory` and keep everything as convoluted as it is now, so I removed the commit.

File: core/src/main/java/org/infinispan/scattered/impl/ScatteredVersionManagerImpl.java [hunk 1]
```diff
@@ -86,9 +84,9 @@
    private AtomicLongArray segmentVersions;
    // holds the topologies in which this node has become the owner of given segment
    private AtomicIntegerArray ownerTopologyIds;
-   private ReadWriteLock scheduledKeysLock = new ReentrantReadWriteLock();
+   private StampedLock scheduledKeysLock = new StampedLock();
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/scattered/impl/ScatteredVersionManagerImpl.java:87]: make `scheduledKeysLock` and `removedKeysLock` final pls :)

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
@danberindei looks good to me. @wburns do you wanna take another look?

**Comment by wburns** [comment]
Seems good to me as far as I can see.

**Comment by wburns** [comment]
@danberindei maybe squash the commits? there are quite a lot

**Comment by danberindei** [comment]
@wburns I thought it would be better to separate them in case we might want to revert one of them later, but sure, I can squash them.

**Comment by danberindei** [comment]
Squashed @wburns

**Comment by wburns** [comment]
Integrated into master, thanks @danberindei !

**Pull request merged**